ksql-parser/src/main/java/io/confluent/ksql/parser/rewrite/StatementRewriteForStruct.java

```
@@ -0,0 +1,109 @@
+import java.util.Objects;
+
+import io.confluent.ksql.parser.tree.DereferenceExpression;
+import io.confluent.ksql.parser.tree.Expression;
+import io.confluent.ksql.parser.tree.FunctionCall;
+import io.confluent.ksql.parser.tree.Node;
+import io.confluent.ksql.parser.tree.QualifiedName;
+import io.confluent.ksql.parser.tree.QualifiedNameReference;
+import io.confluent.ksql.parser.tree.Statement;
+import io.confluent.ksql.parser.tree.StringLiteral;
+import io.confluent.ksql.util.DataSourceExtractor;
+
+public class StatementRewriteForStruct {
+
+  private final Statement statement;
+
+  public StatementRewriteForStruct(
+      final Statement statement,
+      final DataSourceExtractor dataSourceExtractor
+  ) {
+    this.statement = Objects.requireNonNull(statement, "statement");
+  }
+
+  public Statement rewriteForStruct() {
+    return (Statement) new RewriteWithStructFieldExtractors().process(statement, null);
+  }
+
+
+  private class RewriteWithStructFieldExtractors extends StatementRewriter {
+
+    @Override
+    protected Node visitDereferenceExpression(
+        final DereferenceExpression node,
+        final Object context
+    ) {
+      return createFetchFunctionNodeIfNeeded(node, context);
+    }
+
+    private Expression createFetchFunctionNodeIfNeeded(
+        final DereferenceExpression dereferenceExpression,
+        final Object context
+    ) {
+      if (dereferenceExpression.getBase() instanceof QualifiedNameReference) {
+        return getNewDereferenceExpression(dereferenceExpression, context);
+      } else if (dereferenceExpression.getBase() instanceof SubscriptExpression) {
+        return new FunctionCall(
+            QualifiedName.of("FETCH_FIELD_FROM_STRUCT"),
+            ImmutableList.of(
+                dereferenceExpression.getBase(),
+                new StringLiteral(dereferenceExpression.getFieldName())
+            ));
+      }
+      return getNewFunctionCall(dereferenceExpression, context);
```

Are we assuming that this is a DeferenceExpression? That may not always be the case. For example, this could be a function call. Say I have a UDF that returns a STRUCT or I use `FETCH_FIELD_FROM_STRUCT` directly. I think its ok to not permit a dereference of the value returned from a function call but we should return an error in that case.

---

The above fix will take care of this too.